goal_src/jak2/engine/gfx/ocean/ocean.gc

```diff
@@ -5,5 +5,1374 @@
+  (rlet ((vf13 :class vf)
+         (vf14 :class vf)
+         (vf25 :class vf)
+         (vf26 :class vf)
+         (vf27 :class vf)
+         (vf28 :class vf)
+         (vf29 :class vf)
+         (vf30 :class vf)
+         (vf31 :class vf)
+         )
+    (let ((v1-0 *math-camera*))
+      (set-vector! (-> *sky-work* fog) (-> v1-0 pfog0) (-> v1-0 fog-min) (-> v1-0 fog-max) 3071.0)
+      (cond
+        ((-> *time-of-day-context* use-camera-other)
+         (.lvf vf31 (&-> v1-0 camera-temp-other vector 0 quad))
+         (.lvf vf30 (&-> v1-0 camera-temp-other vector 1 quad))
+         (.lvf vf29 (&-> v1-0 camera-temp-other vector 2 quad))
+         (.lvf vf28 (&-> v1-0 camera-temp-other trans quad))
+         (.mov a0-4 vf28)
+         )
+        (else
+          (.lvf vf31 (&-> v1-0 camera-temp vector 0 quad))
+          (.lvf vf30 (&-> v1-0 camera-temp vector 1 quad))
+          (.lvf vf29 (&-> v1-0 camera-temp vector 2 quad))
+          (.lvf vf28 (&-> v1-0 camera-temp trans quad))
+          (.mov a0-5 vf28)
+          )
+        )
+      (.lvf vf26 (&-> v1-0 inv-hmge-scale quad))
+      (.lvf vf14 (&-> v1-0 hmge-scale quad))
+      (.lvf vf25 (&-> v1-0 hvdf-off quad))
+      )
+    (.lvf vf13 (&-> *sky-work* fog quad))
+    (.lvf vf27 (&-> *sky-work* giftag-ocean quad))
+    (.mov v1-3 vf27)
+    0
+    (none)
+    )
+  )
+
+;; ERROR: function was not converted to expressions. Cannot decompile.
+
+;; ERROR: function was not converted to expressions. Cannot decompile.
+
+(defmethod add-colors! ocean ((obj ocean) (arg0 vector) (arg1 ocean-vertex))
+  (let ((s3-0 (new 'stack-no-clear 'vector))
+        (s4-0 (-> obj haze-lights))
+        )
+    (set! (-> s3-0 quad) (-> arg1 pos quad))
+    (set! (-> s3-0 y) 0.0)
+    (vector-normalize! s3-0 1.0)
+    (set! (-> s3-0 y) (+ 0.1 (-> s3-0 y)))
+    (vector-normalize! s3-0 1.0)
+    (let ((s5-0 (new 'stack-no-clear 'vector4)))
+      (let* ((f0-4 (vector-dot (-> s4-0 sun0-normal) s3-0))
+             (f1-2 (vector-dot (-> s4-0 sun1-normal) s3-0))
+             (f2-1 (vector-dot (-> s4-0 moon-normal) s3-0))
+             (f0-5 (fmax 0.0 f0-4))
+             (f30-0 (fmax 0.0 f1-2))
+             (f28-0 (fmax 0.0 f2-1))
+             )
+        (set! (-> s5-0 quad) (-> s4-0 ambi-color quad))
+        (vector4-madd! s5-0 s5-0 (the-as vector4 (-> s4-0 sun0-color)) f0-5)
+        (vector4-madd! s5-0 s5-0 (the-as vector4 (-> s4-0 sun1-color)) f30-0)
+        (vector4-madd! s5-0 s5-0 (the-as vector4 (-> s4-0 moon-color)) f28-0)
+        )
+      (vector4-scale! s5-0 s5-0 128.0)
+      (set! (-> arg0 x) (fmax 0.0 (fmin 255.0 (-> s5-0 x))))
+      (set! (-> arg0 y) (fmax 0.0 (fmin 255.0 (-> s5-0 y))))
+      (set! (-> arg0 z) (fmax 0.0 (fmin 255.0 (-> s5-0 z))))
+      )
+    )
+  0
+  (none)
+  )
+
+(defmethod ocean-method-60 ocean ((obj ocean) (arg0 dma-buffer))
+  (local-vars (sv-48 float) (sv-52 vector) (sv-56 vector))
+  (let ((vertices (the-as (inline-array ocean-vertex) (+ #x70000000 0))))
```

Comment:
this line should be patched to
```
(let ((vertices (scratchpad-object (inline-array ocean-vertex))))
```
The `scratchpad-object` macro should replace any uses of `#x7000000` as an address - this was the hard-coded scratchpad on ps2.